src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Doctrine\QueryCollectionExtension;
+
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\ContextAwareQueryCollectionExtensionInterface;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Bundle\ApiBundle\Context\UserContextInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+
+/** @experimental */
+final class ProductsByTaxonExtension implements ContextAwareQueryCollectionExtensionInterface
+{
+    public function __construct(
+        private UserContextInterface $userContext,
+    ) {
+    }
+
+    public function applyToCollection(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
+        string $operationName = null,
+        array $context = [],
+    ): void {
+        if (!is_a($resourceClass, ProductInterface::class, true)) {
+            return;
+        }
+
+        $user = $this->userContext->getUser();
+        if ($user !== null && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        if (isset($context['filters'])) {
+            $taxonCode = $context['filters']['productTaxons.taxon.code'] ?? null;
+            $this->addSortingToQuery($queryBuilder, $taxonCode, $queryNameGenerator);
+        }
```

Comment:
I think you can make `$taxonCode` not nullable in `addSortingToQuery` method and move the condition with early return to the `applyToCollection`:
```suggestion
        $taxonCode = $context['filters']['productTaxons.taxon.code'] ?? null;
        if (null === $taxonCode) {
            return;
        }
        
        $this->addSortingToQuery($queryBuilder, $taxonCode, $queryNameGenerator);
```